File: internal/handler/opml.go
package handler

import (
	"context"
	"encoding/xml"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strings"
	"time"

	"github.com/kontrolplane/feed/internal/store"
)

// OPML structures

type opml struct {
	XMLName xml.Name `xml:"opml"`
	Version string   `xml:"version,attr"`
	Head    opmlHead `xml:"head"`
	Body    opmlBody `xml:"body"`
}

type opmlHead struct {
	Title       string `xml:"title"`
	DateCreated string `xml:"dateCreated,omitempty"`
}

type opmlBody struct {
	Outlines []opmlOutline `xml:"outline"`
}

type opmlOutline struct {
	Text     string        `xml:"text,attr"`
	Title    string        `xml:"title,attr,omitempty"`
	Type     string        `xml:"type,attr,omitempty"`
	XMLURL   string        `xml:"xmlUrl,attr,omitempty"`
	HTMLURL  string        `xml:"htmlUrl,attr,omitempty"`
	Outlines []opmlOutline `xml:"outline,omitempty"`
}

func (h *Handler) handleExportOPML(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	folders, _ := h.store.Folders(ctx)
	feeds, _ := h.store.Feeds(ctx)

	doc := opml{
		Version: "2.0",
		Head: opmlHead{
			Title:       "kontrolplane/feed",
			DateCreated: time.Now().UTC().Format(time.RFC1123Z),
		},
	}

	for _, folder := range folders {
		group := opmlOutline{Text: folder.Label, Title: folder.Label}
		for _, feed := range feeds {
			if feed.Folder == folder.ID {
				group.Outlines = append(group.Outlines, opmlOutline{
					Text:    feed.Title,
					Title:   feed.Title,
					Type:    "rss",
					XMLURL:  feed.URL,
					HTMLURL: feed.SiteURL,
				})
			}
		}
		if len(group.Outlines) > 0 {
			doc.Body.Outlines = append(doc.Body.Outlines, group)
		}
	}

	w.Header().Set("Content-Type", "application/xml")
	w.Header().Set("Content-Disposition", `attachment; filename="feeds.opml"`)
	w.Write([]byte(xml.Header))
	enc := xml.NewEncoder(w)
	enc.Indent("", "  ")
	enc.Encode(doc)
}

func (h *Handler) handleImportOPML(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	file, _, err := r.FormFile("file")
	if err != nil {
		h.logger.Error("import opml: no file", slog.Any("error", err))
		http.Error(w, "no file uploaded", http.StatusBadRequest)
		return
	}
	defer file.Close()

	data, err := io.ReadAll(io.LimitReader(file, 10<<20)) // 10MB limit
	if err != nil {
		h.logger.Error("import opml: read error", slog.Any("error", err))
		http.Error(w, "failed to read file", http.StatusBadRequest)
		return
	}

	var doc opml
	if err := xml.Unmarshal(data, &doc); err != nil {
		h.logger.Error("import opml: parse error", slog.Any("error", err))
		http.Error(w, "invalid OPML file", http.StatusBadRequest)
		return
	}

	var imported int
	folderCount, _ := h.store.FolderCount(ctx)

	for _, outline := range doc.Body.Outlines {
		if outline.XMLURL != "" {
			// Top-level feed (no folder grouping)
			imported += h.importFeed(ctx, outline, "default")
		} else {
			// Folder with nested feeds
			folderID := strings.ToLower(strings.ReplaceAll(outline.Text, " ", "-"))
			if folderID == "" {
				folderID = "default"
			}
			folderLabel := outline.Text
			if folderLabel == "" {
				folderLabel = "default"
			}
			h.store.UpsertFolder(ctx, store.Folder{ID: folderID, Label: folderLabel}, folderCount)
			folderCount++

			for _, child := range outline.Outlines {
				imported += h.importFeed(ctx, child, folderID)
			}
		}
	}

	h.logger.Info("imported opml", slog.Int("feeds", imported))

	// Return the settings page with a refresh
	if r.Header.Get("HX-Request") == "true" {
		w.Header().Set("HX-Redirect", "/settings")
		w.WriteHeader(http.StatusOK)
		fmt.Fprintf(w, "imported %d feeds", imported)
	} else {
		http.Redirect(w, r, "/settings", http.StatusSeeOther)
	}
}

func (h *Handler) importFeed(ctx context.Context, outline opmlOutline, folderID string) int {
	if outline.XMLURL == "" {
		return 0
	}

	title := outline.Title
	if title == "" {
		title = outline.Text
	}
	if title == "" {
		title = outline.XMLURL
	}

	id := strings.ReplaceAll(strings.TrimPrefix(strings.TrimPrefix(outline.XMLURL, "https://"), "http://"), "/", "-")
	if len(id) > 32 {
		id = id[:32]
	}

	h.store.AddFeed(ctx, store.Feed{
		ID:      id,
		Title:   title,
		URL:     outline.XMLURL,
		Folder:  folderID,
		SiteURL: outline.HTMLURL,
	})

	return 1
}

File: src/opml.ts
import type { Store } from "./store";

export function esc(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
}

function attr(s: string): string | null {
  if (s === "") return null;
  const m = /^="([^"]*)"|^='([^']*)'|^=([^\s>]+)/.exec(s);
  if (!m) return "";
  return m[1] ?? m[2] ?? m[3] ?? "";
}

export function parseAttrs(tag: string): Record<string, string> {
  const out: Record<string, string> = {};
  const re = /([\w:.-]+)(\s*=\s*("[^"]*"|'[^']*'|[^\s>]+))?/g;
  let m: RegExpExecArray | null;
  while ((m = re.exec(tag)) !== null) {
    const name = m[1].toLowerCase();
    if (name === "outline") continue;
    if (m[2] === undefined) {
      out[name] = "";
    } else {
      const v = attr(m[2].trim());
      if (v !== null) out[name] = v;
    }
  }
  return out;
}

export interface OpmlOutline {
  text: string;
  title: string;
  xmlUrl: string;
  htmlUrl: string;
  children: OpmlOutline[];
}

export function parseOpmlOutlines(xml: string): OpmlOutline[] {
  const bodyMatch = /<body[^>]*>([\s\S]*)<\/body\s*>/i.exec(xml);
  const body = bodyMatch ? bodyMatch[1] : xml;
  const roots: OpmlOutline[] = [];
  const stack: OpmlOutline[][] = [roots];
  const nodeStack: OpmlOutline[] = [];
  const re = /<(\/?)outline\b([^>]*?)(\/?)>/gi;
  let m: RegExpExecArray | null;
  while ((m = re.exec(body)) !== null) {
    const closing = m[1] === "/";
    const attrs = parseAttrs(m[2] ?? "");
    const selfClosing = m[3] === "/";
    if (closing) {
      if (stack.length > 1) stack.pop();
      nodeStack.pop();
      continue;
    }
    const node: OpmlOutline = {
      text: attrs.text ?? "",
      title: attrs.title ?? "",
      xmlUrl: attrs.xmlurl ?? "",
      htmlUrl: attrs.htmlurl ?? "",
      children: [],
    };
    stack[stack.length - 1].push(node);
    if (!selfClosing) {
      stack.push(node.children);
      nodeStack.push(node);
    }
  }
  return roots;
}

export function feedIdForUrl(feedUrl: string): string {
  let id = feedUrl.replace(/^https?:\/\//, "").replace(/\//g, "-");
  if (id.length > 32) id = id.slice(0, 32);
  return id;
}

function importOutline(store: Store, o: OpmlOutline, folderId: string): number {
  if (!o.xmlUrl) return 0;
  const title = o.title || o.text || o.xmlUrl;
  store.addFeed({ id: feedIdForUrl(o.xmlUrl), title, url: o.xmlUrl, folder: folderId, siteUrl: o.htmlUrl });
  return 1;
}

export function importOpml(store: Store, xml: string): number {
  const outlines = parseOpmlOutlines(xml);
  let imported = 0;
  let folderCount = store.folderCount();
  for (const o of outlines) {
    if (o.xmlUrl) {
      store.upsertFolder({ id: "default", label: "default" }, folderCount);
      imported += importOutline(store, o, "default");
    } else {
      const folderId = o.text ? o.text.toLowerCase().replaceAll(" ", "-") : "default";
      const label = o.text || "default";
      store.upsertFolder({ id: folderId, label }, folderCount);
      folderCount++;
      for (const child of o.children) imported += importOutline(store, child, folderId);
    }
  }
  return imported;
}

export async function importOpmlFileAsync(store: Store, path: string): Promise<number> {
  const text = await Bun.file(path).text();
  return importOpml(store, text);
}

export function exportOpml(store: Store): string {
  const folders = store.folders();
  const feeds = store.feeds();
  const now = new Date().toUTCString();
  let out = `<?xml version="1.0" encoding="UTF-8"?>\n<opml version="2.0">\n  <head>\n    <title>kontrolplane/feed</title>\n    <dateCreated>${esc(now)}</dateCreated>\n  </head>\n  <body>\n`;
  for (const folder of folders) {
    const group = feeds.filter((f) => f.folder === folder.id);
    if (group.length === 0) continue;
    out += `    <outline text="${esc(folder.label)}" title="${esc(folder.label)}">\n`;
    for (const feed of group) {
      out += `      <outline text="${esc(feed.title)}" title="${esc(feed.title)}" type="rss" xmlUrl="${esc(feed.url)}" htmlUrl="${esc(feed.siteUrl)}"/>\n`;
    }
    out += `    </outline>\n`;
  }
  out += `  </body>\n</opml>\n`;
  return out;
}
